### ISSUE TITLE
Analysis of nested OrderedDicts is very slow

### ISSUE THREAD

**Issue by smphhh** [microsoft/pylance-release#767]
## Environment data

-   Language Server version: 2020.12.2
-   OS and version: MacOS 10.15.5
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.8 via pyenv

## Expected behaviour

Analysis should complete very quickly after making edits to the file

## Actual behaviour

In a VS Code project with a single seemingly simple Python file (see below) and a clean virtualenv with no dependencies any edits to the file cause the Pylance analysis to take multiple seconds.

## Logs

I believe the following lines are relevant:
```
Info  - 10:53:06 AM] [BG(1)] Long operation: checking: /xxx/pylance-test/main.py (3945ms)
[Info  - 10:53:06 AM] [BG(1)] Long operation: analyzing: /xxx/pylance-test/main.py (3945ms)
```

## Code Snippet / Additional information

```python
from collections import OrderedDict

a = OrderedDict([
    ("c1", OrderedDict([
        ("c2", [
            OrderedDict([
                ("key", "k1"),
                ("name", "v1"),
                ("c3", [
                    OrderedDict([
                        ("c4", [
                            OrderedDict([
                                (
                                    "key",
                                    "k2",
                                ),
                                (
                                    "value",
                                    "v2",
                                ),
                                (
                                    "language",
                                    "foo",
                                ),
                                (
                                    "startDate",
                                    "1800-01-01T00:00:00.000+02:00",
                                ),
                            ]),
                        ])
                    ])
                ]),
            ])
        ]),
    ]))
])
```

**Comment by erictraut** [comment]
Thanks for the bug report. Although it's not obvious from the bug title, I'm pretty sure this is the same issue as https://github.com/microsoft/pylance-release/issues/590#issuecomment-748167003, which I'm actively working on. Let's leave this one open until I confirm that they're both the same underlying problem.

**Comment by erictraut** [comment]
I've added some additional internal caching to handle deeply-nested expressions of this sort. The above example previously took about 2800ms to analyze. With the new caching in place, it's now down to about 500ms. This fix will be in the next release of Pylance.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.1.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202110-6-january-2021
